### PR TITLE
Remove anonymous PublicationContext implementation

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today the `PublicationContext` interface has a single anonymous
implementation, and `PublicationTransportHandler` has various methods
that take the variables that this anonymous class captures. This commit
refactors this into a proper class with proper fields and moves the
relevant methods onto this class.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java [hunk 1]
```diff
@@ -406,4 +198,216 @@ private PublishWithJoinResponse acceptState(ClusterState incomingState) {
         }
         return handlePublishRequest.apply(new PublishRequest(incomingState));
     }
+
+    public PublicationContext newPublicationContext(ClusterChangedEvent clusterChangedEvent) {
+        final PublicationContext publicationContext = new PublicationContext(clusterChangedEvent);
+
+        // Build the serializations we expect to need now, early in the process, so that an error during serialization fails the publication
+        // straight away. This isn't watertight since we send diffs on a best-effort basis and may fall back to sending a full state (and
+        // therefore serializing it) if the diff-based publication fails.
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java:207]: Reworded comment

File: server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java [hunk 2]
```diff
@@ -406,4 +198,216 @@ private PublishWithJoinResponse acceptState(ClusterState incomingState) {
...
+
+    private static BytesReference serializeFullClusterState(ClusterState clusterState, Version nodeVersion) throws IOException {
+        final BytesStreamOutput bStream = new BytesStreamOutput();
+        try (StreamOutput stream = CompressorFactory.COMPRESSOR.streamOutput(bStream)) {
+            stream.setVersion(nodeVersion);
+            stream.writeBoolean(true);
+            clusterState.writeTo(stream);
+        }
+        final BytesReference serializedState = bStream.bytes();
+        logger.trace("serialized full cluster state version [{}] for node version [{}] with size [{}]",
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java:220]: Added a TRACE log to record the size of the state here.

File: server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java [hunk 3]
```diff
@@ -406,4 +198,216 @@ private PublishWithJoinResponse acceptState(ClusterState incomingState) {
...
+                        }
+                    } else {
+                        // will send a diff
+                        if (diff == null) {
+                            diff = newState.diff(previousState);
+                        }
+                        if (serializedDiffs.containsKey(node.getVersion()) == false) {
+                            final BytesReference serializedDiff = serializeDiffClusterState(diff, node.getVersion());
+                            serializedDiffs.put(node.getVersion(), serializedDiff);
+                            logger.trace("serialized cluster state diff for version [{}] in for node version [{}] with size [{}]",
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java:272]: Added a TRACE log to record the size of the state here.

**Review by DaveCTurner — COMMENTED** [review]
Git has given up on computing a useful diff here but this was mostly a mechanical process with few meaningful changes. I've added a couple more log lines (noted below) and renamed some parameters especially `responseActionListener` to `listener`.

**Review by original-brownbear — APPROVED** [review]
LGTM, nice cleanup.

Maybe while you're at it, make the fields `lastSeenClusterState` and `currentPublishRequestToSelf` `final` (just add the keyword, they are already effectively final)? It's always weird to refer to a non-final field in a nested class, especially when it's something concurrent like `AtomicReference` (IMO).

**Pull request merged**